src/ssl.c

```diff
@@ -258,6 +260,533 @@ const WOLF_EC_NIST_NAME kNistCurves[] = {
 };
 #endif
 
+#if defined(HAVE_ECH)
+int wolfSSL_CTX_GenerateEchConfig(WOLFSSL_CTX* ctx, const char* publicName)
+{
+    int ret = 0;
+    int rngRet = 0;
+    word16 encLen = DHKEM_X25519_ENC_LEN;
+#ifdef WOLFSSL_SMALL_STACK
+    Hpke* hpke;
+    WC_RNG* rng;
+#else
+    Hpke hpke[1];
+    WC_RNG rng[1];
+#endif
+
+    if (ctx == NULL || publicName == NULL)
+      return BAD_FUNC_ARG;
+
+    ctx->echConfigs = (WOLFSSL_EchConfig*)XMALLOC(sizeof(WOLFSSL_EchConfig),
+        ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    ForceZero(ctx->echConfigs, sizeof(WOLFSSL_EchConfig));
+
+    if (ctx->echConfigs == NULL)
+        return MEMORY_E;
+
+#ifdef WOLFSSL_SMALL_STACK
+    hpke = (Hpke*)XMALLOC(sizeof(Hpke), ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (hpke == NULL)
+        return MEMORY_E;
+
+    rng = (WC_RNG*)XMALLOC(sizeof(Hpke), ctx->heap, DYNAMIC_TYPE_RNG);
+
+    if (rng == NULL)
+    {
+        XFREE(hpke, ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        return MEMORY_E;
+    }
+#endif
+
+    ret = rngRet = wc_InitRng(rng);
+
+    /* set random config id */
+    if (rngRet == 0)
+        ret = wc_RNG_GenerateByte(rng, &ctx->echConfigs->configId);
+
+    if (ret == 0) {
+        /* set the kem id */
+        ctx->echConfigs->kemId = DHKEM_X25519_HKDF_SHA256;
+
+        /* set the cipher suite, only 1 for now */
+        ctx->echConfigs->numCipherSuites = 1;
+        ctx->echConfigs->cipherSuites =
+            (EchCipherSuite*)XMALLOC(sizeof(EchCipherSuite), ctx->heap,
+            DYNAMIC_TYPE_TMP_BUFFER);
+
+        if (ctx->echConfigs->cipherSuites == NULL) {
+            ret = MEMORY_E;
+        }
+        else {
+            ctx->echConfigs->cipherSuites[0].kdfId = HKDF_SHA256;
+            ctx->echConfigs->cipherSuites[0].aeadId = HPKE_AES_128_GCM;
+        }
+    }
+
+    if (ret == 0)
+        ret = wc_HpkeInit(hpke, DHKEM_X25519_HKDF_SHA256, HKDF_SHA256,
+            HPKE_AES_128_GCM, ctx->heap);
+
+    /* generate the receiver private key */
+    if (rngRet == 0 && ret == 0)
```

Comment:
Can't see the need to check rngRe here.

Reply:
because the rngInit can fail but the statements after it can succeed and override ret

Reply:
wait never mind I forgot I also set ret so they won't run if rngRet is nonzero